### ISSUE TITLE
Allow user defined audio file/tones for proximity notification

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#7725]
Currently the proximity notification uses a default system tone only. The volume and noticability will depend on the device and the surrounding noise.

Allow a user defined audio file instead (e.g. from installed ringtones) via chooser to be used as proximity notification.

**Comment by Oskar1983** [comment]
Additional it would be nice to have an option to send a notification to the android notification interface. So you could be able to get the notifications via a fitness tracker, smartwatch, ... 
You could keep smartphone the smartphone in a pocket silently(!) until the notification pops up.

**Comment by moving-bits** [comment]
> Additional it would be nice to have an option to send a notification to the android notification interface. So you could be able to get the notifications via a fitness tracker, smartwatch, ... You could keep smartphone the smartphone in a pocket silently(!) until the notification pops up.

This is supported meanwhile, just set notification type to "text only".

But as c:geo does not support a so-called `ForegroundService` this will only work as long as the screen is on. (There's a separate issue for that.)

**Comment by moving-bits** [comment]
@eddiemuc 
I'm trying to make an audio file selectable using `PersistableUri`, but even on successful selection the permission does not get persisted. Any ideas?

- `SettingsActivity` holds a new `ContentStorageActivityHelper` created for this.
- `PreferenceMapContentBehaviorFragment` has a new preference with a onClick listener which starts file selection using the new CSAH. 
- `SettingsActivity.onActivityResult` calls `CSAH.onActivityResult()`.
- the callback in the CSAH gets called

But neither does the preference key hold the URI nor is it listed under `PersistedDocuments` in system information (stays `null`).

What am I missing here?

I've posted the draft PR #14387. Can you have a look at it? Thanks.

Is there a way to "unselect" a file (eg: if the user wants to return to built-in tone), or do I have to make some UI for letting the user decide between built-in and user-selected audio?

**Comment by eddiemuc** [comment]
Hi @moving-bits ,

you use the CSAH method "selectFile". This method is not meant to create persistent Uris, it is meant for "transient" file selection (e.g. when one selects an image to put on a log file).

To create a persistent Uri, you need to use the CSAH  method "selectPersistableUri". Before that you need to create an enum entry in enum class "PersistableUri" where you also have to specify in which config key the routines will store the persisted Uri etc. I created this back in the days for the "track file" which at the time was the only thing where we needed Persisted Uris for single files - it seems this cause is gone now.

To retrieve a persisted Uri (e.g. if you want to open the file behind it), use the "getUri()" method of PersistedUri.

There's no way currently to "remove" a persistent Uri in the code as far as I know. One could use ContentStorage's "setPersistedDocumentUri" method for this I guess. But the overall logic of whether the user wants a build-on audio or his user-selected audio is use case specific and should stay well out of the ContentStorage/CSAH logic! 

Side note: please don't create a second CSAH instance in SettingsActivity for this. SettingsActivity already has a CSAH instance called "contentStorageHelper", you don't need a second one for a second file selection purpose.

**Comment by moving-bits** [comment]
Thanks, @eddiemuc, for the detailed reply. I'll dive into it.

> you use the CSAH method "selectFile". This method is not meant to create persistent Uris, it is meant for "transient" file selection (e.g. when one selects an image to put on a log file).

ah, I had reused this part from current track file selection which indeed does not persist the uri, hadn't cross-checked with the old approach we used when a single track file uri got persisted.

> To create a persistent Uri, you need to use the CSAH method "selectPersistableUri". Before that you need to create an enum entry in enum class "PersistableUri" where you also have to specify in which config key the routines will store the persisted Uri etc.

this part I have done, but it did not come into effect due to me using `selectFile` instead of `selectPersistableUri`

> I created this back in the days for the "track file" which at the time was the only thing where we needed Persisted Uris for single files - it seems this cause is gone now.

yes, it is, since we have switched from persisting a single track file uri to selecting (and copying!) multiple track files. But I'm reusing that very part for notification files now.

> There's no way currently to "remove" a persistent Uri in the code as far as I know. One could use ContentStorage's "setPersistedDocumentUri" method for this I guess. 

I'll look into it

> But the overall logic of whether the user wants a build-on audio or his user-selected audio is use case specific and should stay well out of the ContentStorage/CSAH logic!

for sure!

> Side note: please don't create a second CSAH instance in SettingsActivity for this. SettingsActivity already has a CSAH instance called "contentStorageHelper", you don't need a second one for a second file selection purpose.

will adapt the PR accordingly

Thanks for the extensive answer, much appreciated!

**Comment by moving-bits** [comment]
One question, though, regarding the "CSAH twice" part:

Those two instances have different purposes, the first one being of `SELECT_FOLDER_PERSISTED` action type for selecting one or more of the configurable folders, the other one being a `SELECT_FILE` action for selecting the audio files, therefore with different follow-up actions. Currently both are created in `onCreate` method of `SettingsActivity`. 

Would it be ok to not create it there but do it "on demand", meaning: when the user selects one of the related actions in one of the fragments? (The user may not switch forth and back between the different selection actions, though, as the CSAH objects would no longer be both available during the whole lifecycle of `SettingsActivity`.) Or better keep both instances?

**Comment by eddiemuc** [comment]
As long as different select-typew lead to different results I think you can use same CSAH instance (and register multiple select-types, one for each purpose).

Also I think there's a separate type for persistent file selection

**Comment by moving-bits** [comment]
Thanks, I adapted the PR accordingly. There's another place where we are using two CSAH in parallel (`RouteTrackUtils`), Separate PR to fix this will follow.

**Comment by moving-bits** [comment]
Implemented with #14387
